### DISCUSSION TITLE
How do I use Explicit HTTPS in reverse proxy mode?

### DISCUSSION THREAD

**Discussion by weibeu** [mitmproxy/mitmproxy#4947] [Q&A] [answered]
I am trying to setup MITM proxy to pass decrypted HTTP body to my backend application. However when I run MITM proxy in reverse mode, I get an error: `Tunnel connection failed: 502 Bad Gateway`.

Here are the options I am passing to mitmproxy:

```sh
$ mitmproxy --listen-port=8080 --listen-host=0.0.0.0 --mode=reverse:http://127.0.0.1:5002 --set=block_global=false --set=ssl_insecure=true
```

What I am expecting here with MITM:
- pretending to client that HTTPS tunnel has been created using HTTP CONNECT method.
- decrypting HTTPS body and passing it to backend application.
- get the response from backend application.
- encrypt the response body and return response to the client.

**Comment by mhils** [comment, marked as answer]
I think you are misunderstanding how reverse proxy mode works. In regular mode you would expect an HTTP CONNECT (https://docs.mitmproxy.org/stable/concepts-howmitmproxyworks/), in reverse proxy mode mitmproxy acts like a normal HTTP(S) server and transparently forwards the connecting to the upstream destination (https://docs.mitmproxy.org/stable/concepts-modes/#reverse-proxy). In other words, in reverse proxy mode, there's no CONNECT necessary or expected as the destination is already known.

You can either 1) change your client to treat the mitmproxy instance as a regular HTTP server or 2) use mitmproxy in regular mode. If you want your client to believe that everything is HTTPS but you need plain HTTP between proxy and server, take a look at https://docs.mitmproxy.org/stable/overview-features/#map-remote.
